### ISSUE TITLE
Unsafe or undocumented assumptions in ldap3.utils.conv.escape_filter_chars

### ISSUE THREAD

**Issue by rtasarz** [cannatag/ldap3#278]
If input data is given as string and it contains backslash character, this function returns its input untouched, that makes it useless to escape anything that may contain backslashes. Also it's pass through behaviour for byte type inputs is rather unintuitive.

**Comment by cannatag** [comment]
Hi, you're right. I must find a better way to check if the passed valued is already escaped. Will work on this for the next release.

**Comment by rtasarz** [comment]
IMO unless package logic on some execution path may call this function over the same input twice, I personally would skip detection, allow already encoded string to be encoded again and leave responsibility to a developer to keep it called only once. I think there's no reliable way to implement such escape detection that wouldn't produce false positives. As nothing more realistic is coming to my mind, let's say searching for windows path `escape_filter_chars('animals\aardvark')` will be detected as escaped but will fail user expectations. And as function is part of public API, someone may have more valid use case for it. Anyway below is code that checks for escaped strings. Of course if you find it useful feel free to do anything you want with it. It assumes, that as I wrote in #279, pretty much all UTF-8 chars are valid ones.

    import re
    def is_filter_safe(text):
        if not type(text) == ((bytes != str) and str or unicode):  # requires str for Py3 or unicode for Py2
              raise ValueError('unicode input expected')
        return (all(c not in text for c in '()*\0') and 
                not re.search('\\\\([^0-9a-fA-F]|(.[^0-9a-fA-F]))', text))

**Comment by cannatag** [comment]
The problem I see is with user input string in a form, so my idea was to not change already escaped strings. But the check can't be reliable, I agree with you on this. Probably the best is leave the burden to detect already escaped strings to the developer. 

May I include your ``is_filter_safe(text)`` code in the utils.conv module?

**Comment by toddjames** [comment]
I'm having the same issue and coincidentally, just discovered this issue on GitHub. I can't offer help with the solution, but I would just like to request that the solution handles the following situation properly:

I am trying to escape a string similar to the following (modified to protect the innocent):

`CN=Smith\, Andrew (Andy),OU=My Users,DC=contoso,DC=com`

This DN was returned by a previous query (specifically, looking at the `member` attribute on a group). I now want to look up a certain attribute for this user. Clearly, the parentheses need to be escaped. I run whatever is returned by the previous query (so, this DN) through `escape_filter_chars()`, which does nothing to the string because it contains a backslash (`\`).

**Comment by cannatag** [comment]
In aware of the problem. In version 2 of ldap3 I added an auto_escape feature that actually performs two different functions: auto encoding and auto escaping. Also both were erroneously applied to all attributes values (where only auto_encoding makes sense). Furthermore escaping were applied to all bytes greater than  0x7f. This is allowed by the RFC buy it's not strictly mandatory for VALID UTF-8 characters.

The escape_filter_chars() function also made two bad assumption: (1) if a slash is present in the filter the text is already escaped, (2) bytes value doesn't need escaping.

This is very far from how the "auto" helpers should work so I'm rewriting this part of the library with the following guidelines:

1. Split auto_encoding from auto_escaping
2. Add a function that check if a text is already escaped
3. Remove both assumption from the escape_filter_chars() function
4. Apply auto_encoding to both the seach filter text and attribute values 
5. Apply auto_escaping only to the search filter text only if it is not already escaped
6. Add the auto_encode parameter to the Connection object (auto_escape is already present)
7. Add an auto_escape parameter to the search() method of the Connection object to override the Connection auto_escape only for the current search.

Will work on this in the weekend. Suggestions appreciated.

**Comment by cannatag** [comment]
I had time to work on this today. Please have a look at the version in dev. It should work as stated in my previous message. Let me know what do you think.

**Comment by rtasarz** [comment]
In my project I've ended with function similar to (simpler, as I'm supporting Py3k only):

    def escape_filter_chars(text, encoding=None, escape=u'\\*()\0'):
        """ Escape chars mentioned in RFC4515. """
        if encoding is None:
            encoding = get_config_parameter('DEFAULT_ENCODING')
        text = to_unicode(text, encoding)
        for char in escape:
            text = text.replace(char, escape_bytes(to_raw(char)))
        return text

As for me it's useful to sometimes pass untouched `*` coming from user forms. Of course feel free to include this version (or any variation of it) in `utils.conv` (same applies to `is_filter_safe(text)`).

**Comment by toddjames** [comment]
@cannatag  thanks for looking at this so quickly. With the latest version in dev, I am getting the following:

```
  File "./adconnection.py", line 40, in get_object
    self._conn.search(self.base_ou, query, attributes=return_attrs)
  File "/home/todd/Projects/al3/env/lib/python3.5/site-packages/ldap3/core/connection.py", line 758, in search
    self.server.schema if self.server else None)
  File "/home/todd/Projects/al3/env/lib/python3.5/site-packages/ldap3/operation/search.py", line 360, in search_operation
    request['filter'] = compile_filter(parse_filter(search_filter, schema, auto_escape, auto_encode).elements[0])  # parse the searchFilter string and compile it starting from the root node
  File "/home/todd/Projects/al3/env/lib/python3.5/site-packages/ldap3/operation/search.py", line 283, in compile_filter
    matching_filter['assertionValue'] = prepare_filter_for_sending(AssertionValue(filter_node.assertion['value']))
  File "/home/todd/Projects/al3/env/lib/python3.5/site-packages/ldap3/protocol/convert.py", line 179, in prepare_filter_for_sending
    ints.append(int(raw_string[i + 1: i + 3], 16))
```

This is with this query:
```
(distinguishedName=CN=Doe\, John A.,OU=My_Users,DC=contoso,DC=com)
```

This happens regardless of whether I use the new escape function, the old one, or none at all.

I added a `print(raw_string[i + 1: i + 3])` before the failing line of code in `convert.py` and it prints `5c` in the Terminal, which I believe is referring to the backslash.

**Comment by cannatag** [comment]
Please check again code in dev. I should have fixed it for your case. Also I've added some specific test cases with backslash in ldap3\test\testSearchOperation.py

**Comment by toddjames** [comment]
Both issues are now fixed for me (the latest one, and the one originally reported in this issue). Many thanks!

**Comment by cannatag** [comment]
I've added checks to encode only well known utf8 syntaxes and types. Will be released in 2.2.

**Comment by cannatag** [comment]
fixed and included in v2.2.0
